### ISSUE TITLE
New compression algorithm for replays

### ISSUE THREAD

**Issue by Wesmania** [FAForever/faf-aio-replayserver#13]
Current format uses zlib and base-64 encodes the archive. This is not ideal, there are better options. Discussions on format choice will happen here.

**Comment by Wesmania** [comment]
Notes from zulip:

```
I did a little compression test just now, zstd gives us 18% smaller replays than zlib with same compression speed
(That's ignoring the 25% we can get by getting rid of base64 encoding)
So if we introduce both, we can convert all existing replays to save 40% or so disk space
LZMA gives us another 16% over zstd actually, so if we don't care about compression speed it might be a better choice
With xz on preset 1, we get 8% more and 2x compression time, on default preset it's 18% and 14x more time
```

tl;dr:

* zstd gives us 18% better compression at same compression speed.
* LZMA gives us an extra 8% over zstd at half the compression speed,
* Or extra 18% at 1/14th compression speed.
* Results may vary. Different compression levels and dictionaries for zstd might yield better results.

I suggest using zstd, plus adding an extra entry in the replay header indicating compression type in case we ever change our mind - for now it'll equal "zstd".

**Comment by Katharsas** [comment]
Could you try LZ4 for comparison?

**Comment by Brutus5000** [comment]
Currently, we have 280GB of replays on production.
New baseline without base64: 210GB (-70GB)
With zstd: ~38gb saved over baseline (total 178 GB).
With LZMA at half speed: additional ~17GB saved (compared to zstd, total 155 GB)
With LZA at 1/14th speed: additional 38GB (total 134 GB).

FAF required 7 (?) years to achieve that amount of data. With zstd that would be ~25GB per year. If we could achieve the savings on existing replays that would last for 4 years until we require the same disk space as now.

Given that CPU is more expensive than storage we should aim for less CPU intensive.

**Comment by Wesmania** [comment]
Okay, we'll go with zstd then. LZ4 would probably be crazy fast, I can't be bothered to benchmark.

**Comment by Wesmania** [comment]
I'll need to prepare a compression dictionary for zstd, but I can just use my own replays for that.

**Comment by Katharsas** [comment]
Will this compression also be active used when client downloads a replay or just for storage?

**Comment by Wesmania** [comment]
What's the difference? If you mean live replays, all traffic is uncompressed.

**Comment by Katharsas** [comment]
Well if you want to decode (or encode) zstd in the java client, you will have to deal with a native JNI dependency, since currently all zsstd implementations seem to be native c code only.

**Comment by Wesmania** [comment]
I'm sure any python compression library is using more than pure python - the only advantage of zlib is that it's a core library. Will a native java dependency hurt us much?

**Comment by Katharsas** [comment]
Depends if it is packaged probably. Usually for a Java dependency that wraps a native binary, you can select which platform the binary should be for (windows, linux etc). But im sure there are some dependencies that just bundle EVERY platform binary inside them, leading to very big installation footprint increase.
Just something to check / keep in mind. I don't think it should be big problem in this case if the Java lib does this properly.

**Comment by Brutus5000** [comment]
Apache Commons Compress has support for Zstandard

**Comment by Katharsas** [comment]
Ok nice
